### PR TITLE
Core Scrabble Completed

### PR DESCRIPTION
**Pull request by davidbrockway**
Core Scrabble Completed - Test Passed

### CODE REVIEW

File: src/scrabble.js [hunk 1]
```diff
@@ -1,5 +1,98 @@
-function scrabble() {
+const scoreValues = {
```
Comment (marc-alzapiedi) [inline src/scrabble.js:1]: I like the use of the object to assign the letters a score. Looking at other peoples code a lot of people took this approach. Its very neat, The way I did it is a bit more obnoxious.

File: src/scrabble.js [hunk 2]
```diff
@@ -1,5 +1,98 @@
...
+  Z: 10,
+}
 
+function isValidInput(word) {
+  return (word !== undefined && word !== null);
 }
 
-module.exports = scrabble
+function scrabble(word) {
+  if (!isValidInput(word)) {
```
Comment (marc-alzapiedi) [inline src/scrabble.js:36]: I had to think the use of the bang on lines 32 and 36 but after thinking it through it makes sense. Clever!

File: src/scrabble.js [hunk 3]
```diff
@@ -1,5 +1,98 @@
...
+    return 0;
+  }
+
+  let s = 0; // stands for score
```
Comment (marc-alzapiedi) [inline src/scrabble.js:40]: Personally I would avoid defining the variable like you have in this way where the variable is a single letter with a comment on what it means. 

It may be better to directly define the variable (score).

File: src/scrabble.js [hunk 4]
```diff
@@ -1,5 +1,98 @@
...
+function scrabble(word) {
+  if (!isValidInput(word)) {
+    return 0;
+  }
+
+  let s = 0; // stands for score 
+  let wordMultiplier = 1;
+  let i = 0;
+
+  while (i < word.length) {
```
Comment (marc-alzapiedi) [inline src/scrabble.js:44]: love the while loop, very clever!

File: src/scrabble.js [hunk 5]
```diff
@@ -1,5 +1,98 @@
...
+    let letter = word[i].toUpperCase();
+
+    if (letter === '{' || letter === '[') {
+      let closingChar = letter === '{' ? '}' : ']';
+      let j = word.indexOf(closingChar, i + 1);
+      if (j === -1) {
```
Comment (marc-alzapiedi) [inline src/scrabble.js:50]: Looking at this if statement seems counterintuitive first glance, but I think I know the trick your using. Also on line 48 it seems like you are checking to see if there are any characters that were included in the function argument that are not accepted, like (\ | % £ $).

File: src/scrabble.js [hunk 6]
```diff
@@ -1,5 +1,98 @@
...
+  K: 5,
+  J: 8,
+  X: 8,
+  Q: 10,
+  Z: 10,
+}
 
+function isValidInput(word) {
+  return (word !== undefined && word !== null);
 }
```
Comment (alexjshaw) [inline src/scrabble.js:33]: nice clean way of exiting the scrabble function if a word isn't entered.

File: src/scrabble.js (hunk 3 above)
Comment (alexjshaw) [inline src/scrabble.js:40]: As Marc said, use more descriptive variable names! You only have to type them the first time, after that you can autofill them so there's no excuse for being lazy!

File: src/scrabble.js [hunk 7]
```diff
@@ -1,5 +1,98 @@
...
+
+  let s = 0; // stands for score 
+  let wordMultiplier = 1;
+  let i = 0;
+
+  while (i < word.length) {
+    let letter = word[i].toUpperCase();
+
+    if (letter === '{' || letter === '[') {
+      let closingChar = letter === '{' ? '}' : ']';
```
Comment (alexjshaw) [inline src/scrabble.js:48]: Could you explain line 48, I don't quite understand what the === '{' ? '}' : ']' part is actually doing.

File: src/scrabble.js [hunk 8]
```diff
@@ -1,5 +1,98 @@
...
+      i++;
+    } else {
+      return 0;
+    }
+  }
+
+  return s * wordMultiplier;
+
+}
+
```
Comment (alexjshaw) [inline src/scrabble.js:94]: This whole scrabble function seems a bit bloated. Breaking it down in to smaller functions would make it easier to see what each part is doing and check if individual sections are working correctly. Would help to isolate problems and tackle one issue at a time rather than trying to fix every issue at the same time.

File: src/scrabble.js [hunk 9]
```diff
@@ -1,5 +1,98 @@
...
+      i = j + 1;
+    } else if (letter === '|') {
+      let j = word.indexOf('|', i + 1);
+      if (j === -1 || (j - i) % 2 === 0) {
+        return 0;
+      }
+      let letters = word.substring(i + 1, j).toUpperCase();
+      let letterMultiplier = 1;
+      for (let k = 0; k < letters.length; k++) {
+        let letterScore = scoreValues[letters[k].toUpperCase()] || 0;
```
Comment (alexjshaw) [inline src/scrabble.js:69]: If you're getting to the point of using k as a variable then you're probably going a bit too deep with the nested loops.

**Review by alexjshaw — COMMENTED** [review]
Good effort overall, some nice logic at points but I think you've confused yourself a bit at times. Loops within loops within loops are always going to be hard to follow, and trying to break some of that out to its own functions might have made it easier to see what was working and what wasn't. Good effort though David.

File: src/scrabble.js (hunk 7 above)
Comment (ManNavic) [inline src/scrabble.js:48]: Have the same question as Alex
